client/rpc-spec-v2/src/chain_head/subscription.rs

```diff
@@ -20,266 +20,1045 @@
+	/// Unpin __was__ called.
+	Unpinned,
+	/// The block was registered by both events (`Finalized` and `BestBlock` events).
+	///
+	/// Unpin __was__ called.
+	FullyUnpinned,
+}
+
+impl BlockStateMachine {
+	fn new() -> Self {
+		BlockStateMachine::Registered
+	}
+
+	fn advance_register(&mut self) {
+		match self {
+			BlockStateMachine::Registered => *self = BlockStateMachine::FullyRegistered,
+			BlockStateMachine::Unpinned => *self = BlockStateMachine::FullyUnpinned,
+			_ => (),
+		}
+	}
+
+	fn advance_unpin(&mut self) {
+		match self {
+			BlockStateMachine::Registered => *self = BlockStateMachine::Unpinned,
+			BlockStateMachine::FullyRegistered => *self = BlockStateMachine::FullyUnpinned,
+			_ => (),
+		}
+	}
+
+	fn was_unpinned(&self) -> bool {
+		match self {
+			BlockStateMachine::Unpinned => true,
+			BlockStateMachine::FullyUnpinned => true,
+			_ => false,
+		}
+	}
+}
+
+struct BlockState {
+	/// The state machine of this block.
+	state_machine: BlockStateMachine,
+	/// The timestamp when the block was inserted.
+	timestamp: Instant,
+}
+
+/// The state of a single subscription ID.
+struct SubscriptionState<Block: BlockT> {
 	/// The `runtime_updates` parameter flag of the subscription.
 	runtime_updates: bool,
 	/// Signals the "Stop" event.
 	tx_stop: Option<oneshot::Sender<()>>,
-	/// The blocks pinned.
-	blocks: HashSet<Block::Hash>,
-	/// The maximum number of pinned blocks allowed per subscription.
-	max_pinned_blocks: usize,
+	/// Track the block hashes available for this subscription.
+	///
+	/// This implementation assumes:
+	/// - most of the time subscriptions keep a few blocks of the chain's head pinned
+	/// - iteration through the blocks happens only when the hard limit is exceeded.
+	///
+	/// Considering the assumption, iterating (in the unlike case) the hashmap O(N) is
+	/// more time efficient and code friendly than paying for:
+	/// - extra space: an extra BTreeMap<Instant, Hash> to older hashes by oldest insertion
+	/// - extra time: O(log(N)) for insert/remove/find each `pin` block time per subscriptions
+	blocks: HashMap<Block::Hash, BlockState>,
+}
+
+impl<Block: BlockT> SubscriptionState<Block> {
+	/// Trigger the stop event for the current subscription.
+	///
+	/// This can happen on internal failure (ie, the pruning deleted the block from memory)
+	/// or if the subscription exceeded the available pinned blocks.
+	fn stop(&mut self) {
+		if let Some(tx_stop) = self.tx_stop.take() {
+			let _ = tx_stop.send(());
+		}
+	}
+
+	/// Keep track of the given block hash for this subscription.
+	///
+	/// This does not handle pinning in the backend.
+	///
+	/// Returns:
+	/// - true if this is the first time that the block is registered
+	/// - false if the block was already registered
+	fn register_block(&mut self, hash: Block::Hash) -> bool {
+		match self.blocks.entry(hash) {
+			Entry::Occupied(mut occupied) => {
+				let block_state = occupied.get_mut();
+
+				block_state.state_machine.advance_register();
```

Comment:
Nit: the name isn't the clearest to me here, but then I wonder in general whether we want to uphold the flow that a block is first registered as "best" and then next as "finalized", and in this case I wonder whether a general `register_block` function is a little vague and it might be better to have:

```rust
fn register_best_block(..)
fn register_finalized_block(..)
```

Just to enforce this ordering a little better (functions can deny if not called in the right order or whatever).

But yeah just a thought :)

Reply:
Ah I see now I've looked down; this is all exposed via `pin_block` and `unpin_block` essentially, and so yeah, I guess the idea is simply that `pin_block` can be called as many times as wanted (but at least twice; once for best and once for Finalized) and then`unpin_block` will ensure the block isn't re-pinned once it's been called (and will actually also remove the block entirely if either of these happens:

- `pin_block()`, `pin_block()`, `unpin_block()`
- `pin_block()`, `unpin_block()`, `pin_block()`

It's a little hard to "kno" what will happen I feel based on the calls, but I get that ultimately there is one pin and one unpin method that we expose, and so things have to just be correct behind that interface re best and finalized blocks